### DISCUSSION TITLE
Support for ephemeral processes

### DISCUSSION THREAD

**Discussion by rupurt** [F1bonacc1/process-compose#61] [Ideas]
Howdy 👋,

Love the project. Do you have any plans to support ephemeral processes for things like batch or cron jobs?

**Comment by F1bonacc1** [comment]
Hi,

Glad you like it :)
Can you elaborate a little bit more about the batch use case?

For cron jobs will something like this work for you:
```yaml
processes:
  cleaner:
    command: "find ./ -type f -mtime +7 -name '*.bak' -execdir rm -- '{}' \\;"
    working_dir: "/somedir"
    availability:
      restart: "always"
      backoff_seconds: 604800 # every 7 days
```
Or you are looking for something that follows closer the cron syntax?

**Reply by rupurt** [reply]
I'd love to use it to elastically scale up/down a group of processes for kafka consumers

**Reply by F1bonacc1** [reply]
It was always on my to-do list.
In cases like this, I prefer to hear from users like you, if this is actually something that will be useful.

I will add this in one of the next releases.

**Reply by rupurt** [reply]
That would be really awesome :pray:

**Reply by thenonameguy** [reply]
@F1bonacc1 I think the use-case of @rupurt could be also solved with the existing replica support (that is fairly undocumented/unexposed).
Adding an RPC mechanism (`process-compose process scale kafka-consumer 3`) to allow for setting the replica count dynamically would allow external "proper" schedulers to deal with these scenarios.

**Reply by rupurt** [reply]
@thenonameguy that sounds pretty much like the API I would want.

**Reply by F1bonacc1** [reply]
Yes that makes sense, I can add that as well. 

Just one think to remember, while the scaling is suitable for consumers, if the scaled process acts as a server, using the same port will be an issue.

**Reply by thenonameguy** [reply]
That is why the 'PC_REPLICA_NUM' env var is passed to the process, right?

**Reply by F1bonacc1** [reply]
Yes

**Reply by rupurt** [reply]
That would be great. If I could dynamically bind to a port it should cover all cases.

**Reply by thenonameguy** [reply]
AFAIK Kafka consumers don't need to bind a port as they don't serve data themselves.

If you have this need you can always wrap your server exec command with a shell script that does a simple addition using a base port + the replica int env var above.

**Reply by rupurt** [reply]
Yeah they don't. I meant in the context of other ephemeral processes like a group of web servers

**Comment by F1bonacc1** [comment]
Hey @rupurt, @thenonameguy,

This became a significant refactoring effort and took much longer than expected.
I am a little bit concerned about releasing it as GA at this point, simply because of all the changes in the underlying logic.
Are you up to checking an RC, if it meets your expectations, if anything is missing or doesn't work as expected?

**Reply by thenonameguy** [reply]
Sure! Just tell me which branch it lives on.

**Reply by F1bonacc1** [reply]
Thank you @thenonameguy!

I published a pre-release:
https://github.com/F1bonacc1/process-compose/releases/tag/v0.55.0

The branch is:
https://github.com/F1bonacc1/process-compose/tree/feature/scale-processes

**Reply by thenonameguy** [reply]
```
  log_location: ./log_file.{PC_LOG}.log  # <- {PC_LOG} will be replaced with replica number. If more than one replica and PC_LOG not specified, the replica number will be contaminated to the file end.
```
Design consideration:
`PC_REPLICA_NUM` already exists in the env vars for this purpose, I think it's mentally easier that it gets reused in this string substitution context.

**Reply by F1bonacc1** [reply]
That's a very good point @thenonameguy. I also considered it at first, the reason I didn't go for it is to emphasize that it's not an environment variable, to prevent someone from using it elsewhere for automatic substitution.
But if you think it's more intuitive, it's an easy change.

**Reply by thenonameguy** [reply]
I think it is better, also the name is more telling.
For consistency maybe also interpolate in the process name env var?

**Reply by F1bonacc1** [reply]
Sure, I will change it to `PC_REPLICA_NUM` .
The process name is calculated on the fly based on the number of replicas. I wanted to keep API BW compatibility + unique process name + didn't want to affect the existing process names for users that don't use the scaling/replicas feature.

**Reply by F1bonacc1** [reply]
@thenonameguy, @rupurt 

If everything else works as expected. I will make this an official release.

**Reply by rupurt** [reply]
Hey @F1bonacc1,

Sorry, I haven't had a chance to test it out yet. Back from vacation so will give it a try this week.

**Reply by thenonameguy** [reply]
Everything else worked for me very well @F1bonacc1 , spent around an hour playing with it, LGTM.

**Comment by dzmitry-lahoda** [comment]
I need cron job. 

I do not want to code it into my app and run it all the time.

I could have run some cron app script all the time, and it will launch my app. So that would not be so easy to see in PC TUI.

In systemd I have timers 

https://akashrajpurohit.com/blog/systemd-timers-vs-cron-jobs/

And I have most degen way - bash with sleep loop for simple case. Will do this. 

So systemd like timers feels interesting.

**Comment by dzmitry-lahoda** [comment]
So `cron jobs` vs `replica sets` -  seems different things.

Replica sets are complicated. In K8S they just work because of full isolation of containers (none in PC). What rules will be for isolation of processes in PC scaled up?

Another issues, K8S has RS because it also have autoscalers, so actually final features is metric based autoscaler. It pretty fills overkill and will never be target of PC. Right?
 
Another point, may you just cp yaml block in muliline editor and make clones and restart. I assume PC is not tool to run server side processes for long (serviced/k8s are). So workaround is easy - make yaml.

**Comment by dzmitry-lahoda** [comment]
So why make YAML is good for scale but not cron ? Visibility in UI. If code bash cron wrapper cannot see in TUI nice view of CRON invocations. So simplicity and speed of rebuild/restart to is what is killer feature of PC, CRON integration for sure adds to that agenda.

**Comment by dzmitry-lahoda** [comment]
For periodic(not cron) can use simple wrap comand like `watch --differences --no-title  --interval=3 --exec foobar`.

As for scale, I guess before scale, it can be found out that in K8S first goes hot reload of config, so it may be that PC to think of how and why hot reload before scale.

K8S has controllers to define dependencies of reload. So what would be rule to reload when upscale or downscale happens?

**Comment by F1bonacc1** [comment]
As a temporary workaround for cron jobs you can consider something like:
```yaml
processes:
  job:
    command: echo Hi
    availability:
      restart: "always"
      backoff_seconds: 604800 # will run once a week
```

**Comment by F1bonacc1** [comment]
Cron Jobs added in v1.87.0
https://f1bonacc1.github.io/process-compose/scheduled-processes/
